### PR TITLE
Optimize async code in model binders

### PR DESCRIPTION
**Pull request by rynowak**
Optimize the 'no-op' path for our model binders to return a cached task
where possible without going async.

### CODE REVIEW

**Comment by rynowak** [comment]
OK, now this is really where the sausage gets made (delicious perf-sausage). We can avoid going async in all of the cases where a model binder doesn't actually bind anything, and this helps us avoid allocating a LOT of tasks.

Model binding still isn't in a great place after this change, but it's a big improvement. I nuked `BindingSourceModelBinder` because it's an anti-pattern with this change. You **need** to care about perf if you're writing a model binder unfortunately.

Before (includes changes in #3104):
image

After:
image


**YOWZA**
This change cuts the number of `Task<ModelBindingResult>` allocations in half. Moreover we avoid allocating a bunch of async state machines we used to have to instantiate - see `BindingSourceModelBinder+<horriblename>` in the first set of results. This change is saving us 8-10mb over 3000 requests.

**Comment by benaadams** [comment]
Now you are cooking with gas! :+1:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/GenericModelBinder.cs [hunk 1]
```diff
@@ -14,9 +14,20 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
...
+                return ModelBindingResult.NoResultAsync;
+            }
+
+            return BindModelCoreAsync(bindingContext, binderType);
+        }
+
+        private async Task<ModelBindingResult> BindModelCoreAsync(ModelBindingContext bindingContext, Type binderType)
+        {
+            binderType = ResolveBinderType(bindingContext);
             if (binderType != null)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/GenericModelBinder.cs]: Why not pass in `binderType` from `BindModelAsync()`? And why does the value need to be `null`-checked again?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/FormCollectionModelBinder.cs [hunk 2]
```diff
@@ -18,7 +18,17 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
...
+            {
+                return ModelBindingResult.NoResultAsync;
+            }
+
+            return BindModelCoreAsync(bindingContext);
+        }
+
+        private async Task<ModelBindingResult> BindModelCoreAsync(ModelBindingContext bindingContext)
         {
             if (bindingContext.ModelType != typeof(IFormCollection))
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/FormCollectionModelBinder.cs]: Why does this block still exist?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/BodyModelBinder.cs [hunk 3]
```diff
@@ -14,19 +14,31 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
...
-        protected async override Task<ModelBindingResult> BindModelCoreAsync(
-            [NotNull] ModelBindingContext bindingContext)
+        /// <summary>
+        /// Attempts to bind the model using formatters.
+        /// </summary>
+        /// <param name="bindingContext">The <see cref="ModelBindingContext"/>.</param>
+        /// <returns>
+        /// A <see cref="Task{ModelBindingResult}"/> which when completed returns a <see cref="ModelBindingResult"/>.
+        /// </returns>
+        protected async Task<ModelBindingResult> BindModelCoreAsync([NotNull] ModelBindingContext bindingContext)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/BodyModelBinder.cs]: `private` ?

**Comment by dougbu** [comment]
Looks as if `TestBindingSourceModelBinder` still exists and that inherits from `BindingSourceModelBinder`. That adds up to 128 functional test failures.

Does our use of `BindingSourceModelBinder` as a basis for a binder in a functional test imply the `abstract` class would be useful to our customers? Or is `TestBindingSourceModelBinder` another anti-pattern -- a binder that should be checking the `BindingSource` match for itself?

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/GenericModelBinder.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/GenericModelBinder.cs]: It might have decided to become null ¯\_(ツ)_/¯

**Comment by davidfowl** [comment]
CI hates this change

**Comment by rynowak** [comment]
> Does our use of BindingSourceModelBinder as a basis for a binder in a functional test imply the abstract class would be useful to our customers? Or is TestBindingSourceModelBinder another anti-pattern -- a binder that should be checking the BindingSource match for itself?

I don't think it's worth keeping. We should nuke it from our functional tests.

**Comment by rynowak** [comment]
> CI hates this change

Good thing I haven't checked it in yet then. Making all the functional tests happy is usually something I do after addressing PR feedback.

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/FormCollectionModelBinder.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/FormCollectionModelBinder.cs]: Cleaned up

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/BodyModelBinder.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/BodyModelBinder.cs]: sure

**Comment by rynowak** [comment]
Updated

**Comment by dougbu** [comment]
:shipit:

**Comment by davidfowl** [comment]
@rynowak This change went in right?

**Comment by rynowak** [comment]
Yep. So many more perfs now.

**Pull request closed without merging**